File: apps/macos/Sources/SJTUCanvasDownloader/Engine/EngineModels.swift
import Foundation

// Wire types of the sjtu-canvas-engine JSON-RPC protocol (engine/src/rpc.rs,
// engine/src/models.rs, engine/src/downloads). EngineClient converts between
// snake_case and camelCase keys.

struct Profile: Decodable, Hashable, Sendable {
    var id: String
    var name: String
    var shortName: String?
    var avatarUrl: String?
}

struct AccountInfo: Decodable, Hashable, Sendable {
    var authenticated = false
    var profile: Profile?
    /// False when the login could not be protected with a keychain key; it
    /// then lasts only until the app quits.
    var persisted = false
    /// Set by account.get with verify: the login was confirmed with Canvas.
    var verified: Bool?
}

struct LoginStatus: Decodable, Hashable, Sendable {
    var attemptId: String
    /// preparing | waiting | reconnecting | authorizing | authorized | expired | cancelled | error
    var state: String
    var message: String
    var generation: Int
    var revision: Int64
    /// The QR code as a base64 PNG while waiting for the scan.
    var qrPng: String?
    var expiresAt: String?
}

struct Course: Decodable, Identifiable, Hashable, Sendable {
    var id: String
    var name: String
    var courseCode: String
    var term: String?
    var teacher: String?
    /// active | invited_or_pending | completed
    var enrollmentState: String

    var isCurrent: Bool { enrollmentState != "completed" }
}

struct CourseList: Decodable, Sendable {
    var courses: [Course]
}

struct Lesson: Decodable, Identifiable, Hashable, Sendable {
    var videoId: String
    var title: String
    var beginTime: String
    var endTime: String
    var classroom: String
    var available: Bool
    /// resource | canvas-lti | historical (课堂视频旧版)
    var source: String?

    var id: String { videoId }
}

struct LessonList: Decodable, Sendable {
    var lessons: [Lesson]
}

struct CourseFile: Decodable, Identifiable, Hashable, Sendable {
    var id: String
    var displayName: String
    var filename: String
    var size: Int64
    var contentType: String?
    var updatedAt: String?

    var name: String { displayName.trimmed.isEmpty ? filename : displayName }
}

struct FileList: Decodable, Sendable {
    var files: [CourseFile]
}

struct TrackSize: Decodable, Hashable, Sendable {
    /// ready | missing | unavailable
    var status: String
    var size: Int64?

    static let unavailable = TrackSize(status: "unavailable", size: nil)
}

struct LessonSizes: Decodable, Sendable {
    var videoId: String
    var tracks: [String: TrackSize]
}

struct DownloadInfo: Decodable, Identifiable, Hashable, Sendable {
    var id: String
    /// video | file
    var kind: String
    var courseId: String
    var courseName: String
    var resourceId: String
    var track: String?
    var title: String
    /// "第 01 讲 · 电脑屏幕" for videos, the file name for course files.
    var displayName: String
    var beginTime: String?
    var destination: String
    /// The final location once known (the file is `<path>.part` until done).
    var filePath: String?
    /// queued | downloading | paused | completed | failed | cancelled
    var status: String
    var received: Int64
    var total: Int64?
    /// Bytes per second while downloading.
    var speed: Double
    var error: String?
    var createdAt: Date
    var updatedAt: Date
    var completedAt: Date?

    var isUnfinished: Bool { status == "queued" || status == "downloading" || status == "paused" }
    var isRunning: Bool { status == "queued" || status == "downloading" }
    var isCompleted: Bool { status == "completed" }
    var isStopped: Bool { status == "failed" || status == "cancelled" }
}

struct DownloadProgress: Decodable, Sendable {
    var id: String
    var received: Int64
    var total: Int64?
    var speed: Double
}

struct DownloadCounts: Decodable, Hashable, Sendable {
    var all = 0
    /// Not finished: queued, downloading or paused.
    var active = 0
    /// Queued or downloading.
    var running = 0
    var completed = 0
    /// Failed or cancelled.
    var failed = 0
}

struct DownloadList: Decodable, Sendable {
    var items: [DownloadInfo]
    var counts: DownloadCounts
}

struct CreateResult: Decodable, Sendable {
    var created: [DownloadInfo]
    var skipped: [SkippedItem]
}

struct SkippedItem: Decodable, Sendable {
    var index: Int
    /// invalid | queued | downloaded
    var reason: String
    var message: String
}

/// One item of downloads.create. Absent optionals stay off the wire: the
/// engine refuses unknown fields.
struct NewDownload: Encodable, Sendable {
    var kind: String
    var courseId: String
    var courseName: String
    var lessonId: String?
    var fileId: String?
    var title: String
    var beginTime: String?
    var track: String?
    var size: Int64?

    static func video(course: Course, lesson: Lesson, track: String, size: Int64?) -> NewDownload {
        NewDownload(
            kind: "video",
            courseId: course.id,
            courseName: course.name,
            lessonId: lesson.videoId,
            title: lesson.title,
            beginTime: lesson.beginTime.trimmed.isEmpty ? nil : lesson.beginTime,
            track: track,
            size: size
        )
    }

    static func file(course: Course, file: CourseFile) -> NewDownload {
        NewDownload(
            kind: "file",
            courseId: course.id,
            courseName: course.name,
            fileId: file.id,
            title: file.name,
            size: file.size > 0 ? file.size : nil
        )
    }
}

struct ProxySettings: Codable, Hashable, Sendable {
    /// system | direct | custom
    var mode = "system"
    var url: String?
}

struct Preferences: Codable, Hashable, Sendable {
    var downloadDir = ""
    var askDestination = false
    var concurrency = 3
    var defaultTracks = ["slides", "teacher"]
    var proxy = ProxySettings()
}

struct SettingsInfo: Decodable, Hashable, Sendable {
    var preferences: Preferences
    var defaultDownloadDir: String
    var concurrencyMax: Int
    var dataDir: String
    var engineVersion: String
    var testMode: Bool
    var fakeSchool: Bool
}

struct InitializeResult: Decodable, Sendable {
    var `protocol`: Int
    var version: String
    var settings: SettingsInfo
    var account: AccountInfo
}

// MARK: - Request parameters

struct NoParams: Encodable, Sendable {}

struct InitializeParams: Encodable, Sendable {
    var sessionKey: String?
    var downloadsFolder: String?
}

struct VerifyParams: Encodable, Sendable {
    var verify: Bool
}

struct CourseParams: Encodable, Sendable {
    var courseId: String
}

struct SizesParams: Encodable, Sendable {
    var courseId: String
    var lessonId: String
    var tracks: [String]
    var refresh: Bool
}

struct CreateParams: Encodable, Sendable {
    var items: [NewDownload]
    var destination: String?
}

struct ListParams: Encodable, Sendable {
    var filter: String?
    var query: String?
    var limit: Int?
}

struct IDParams: Encodable, Sendable {
    var id: String
}

struct SettingsUpdateParams: Encodable, Sendable {
    var preferences: Preferences
}

struct Empty: Decodable, Sendable {}

struct CountResult: Decodable, Sendable {
    var count: Int
}

File: apps/windows/SJTUCanvasDownloader/Services/EngineModels.cs
using System.Text.Json.Serialization;

namespace CanvasDownloader.Services;

// Wire types of the sjtu-canvas-engine JSON-RPC protocol (engine/src/rpc.rs).
// Property names map to snake_case through the client's naming policy.

public sealed class Profile
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string ShortName { get; set; } = "";
    public string? AvatarUrl { get; set; }
}

public sealed class AccountInfo
{
    public bool Authenticated { get; set; }
    public Profile? Profile { get; set; }
    /// <summary>False when the login could not be protected with a Credential Manager key.</summary>
    public bool Persisted { get; set; }
    public bool Verified { get; set; }
}

public sealed class LoginStatus
{
    public string AttemptId { get; set; } = "";
    /// <summary>preparing | waiting | reconnecting | authorizing | authorized | expired | cancelled | error</summary>
    public string State { get; set; } = "";
    public string Message { get; set; } = "";
    public int Generation { get; set; }
    public long Revision { get; set; }
    /// <summary>The QR code as a base64 PNG while waiting for the scan.</summary>
    public string? QrPng { get; set; }
    public DateTimeOffset? ExpiresAt { get; set; }
}

public sealed class Course
{
    public string Id { get; set; } = "";
    public string Name { get; set; } = "";
    public string CourseCode { get; set; } = "";
    public string? Term { get; set; }
    public string? Teacher { get; set; }
    /// <summary>active | invited_or_pending | completed</summary>
    public string EnrollmentState { get; set; } = "";
}

public sealed class CourseList
{
    public List<Course> Courses { get; set; } = [];
}

public sealed class Lesson
{
    public string VideoId { get; set; } = "";
    public string Title { get; set; } = "";
    public string BeginTime { get; set; } = "";
    public string EndTime { get; set; } = "";
    public string Classroom { get; set; } = "";
    public bool Available { get; set; }
    /// <summary>resource | canvas-lti | historical (课堂视频旧版)</summary>
    public string? Source { get; set; }
}

public sealed class LessonList
{
    public List<Lesson> Lessons { get; set; } = [];
}

public sealed class CourseFile
{
    public string Id { get; set; } = "";
    public string DisplayName { get; set; } = "";
    public string Filename { get; set; } = "";
    public long Size { get; set; }
    public string? ContentType { get; set; }
    public DateTimeOffset? UpdatedAt { get; set; }
}

public sealed class FileList
{
    public List<CourseFile> Files { get; set; } = [];
}

public sealed class TrackSize
{
    /// <summary>ready | missing | unavailable</summary>
    public string Status { get; set; } = "";
    public long? Size { get; set; }
}

public sealed class LessonSizes
{
    public string VideoId { get; set; } = "";
    public Dictionary<string, TrackSize> Tracks { get; set; } = [];
}

public sealed class DownloadInfo
{
    public string Id { get; set; } = "";
    /// <summary>video | file</summary>
    public string Kind { get; set; } = "";
    public string CourseId { get; set; } = "";
    public string CourseName { get; set; } = "";
    public string ResourceId { get; set; } = "";
    public string? Track { get; set; }
    public string Title { get; set; } = "";
    public string DisplayName { get; set; } = "";
    public string? BeginTime { get; set; }
    public string Destination { get; set; } = "";
    public string? FilePath { get; set; }
    /// <summary>queued | downloading | paused | completed | failed | cancelled</summary>
    public string Status { get; set; } = "";
    public long Received { get; set; }
    public long? Total { get; set; }
    public double Speed { get; set; }
    public string? Error { get; set; }
    public DateTimeOffset CreatedAt { get; set; }
    public DateTimeOffset? CompletedAt { get; set; }

    [JsonIgnore]
    public bool IsUnfinished => Status is "queued" or "downloading" or "paused";

    [JsonIgnore]
    public bool IsStopped => Status is "failed" or "cancelled";
}

public sealed class DownloadProgress
{
    public string Id { get; set; } = "";
    public long Received { get; set; }
    public long? Total { get; set; }
    public double Speed { get; set; }
}

public sealed class DownloadCounts
{
    public long All { get; set; }
    public long Active { get; set; }
    public long Running { get; set; }
    public long Completed { get; set; }
    public long Failed { get; set; }
}

public sealed class DownloadList
{
    public List<DownloadInfo> Items { get; set; } = [];
    public DownloadCounts Counts { get; set; } = new();
}

public sealed class CreateResult
{
    public List<DownloadInfo> Created { get; set; } = [];
    public List<SkippedItem> Skipped { get; set; } = [];
}

public sealed class SkippedItem
{
    public int Index { get; set; }

    /// <summary>invalid | queued | downloaded</summary>
    public string Reason { get; set; } = "";

    public string Message { get; set; } = "";
}

/// <summary>One item of downloads.create; unused fields must stay off the wire.</summary>
public sealed class NewDownload
{
    public string Kind { get; set; } = "video";
    public string CourseId { get; set; } = "";
    public string CourseName { get; set; } = "";

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? LessonId { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? FileId { get; set; }

    public string Title { get; set; } = "";

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? BeginTime { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Track { get; set; }

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public long? Size { get; set; }
}

public sealed class ProxySettings
{
    /// <summary>system | direct | custom</summary>
    public string Mode { get; set; } = "system";

    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public string? Url { get; set; }
}

public sealed class Preferences
{
    public string DownloadDir { get; set; } = "";
    public bool AskDestination { get; set; }
    public int Concurrency { get; set; } = 3;
    public List<string> DefaultTracks { get; set; } = ["slides", "teacher"];
    public ProxySettings Proxy { get; set; } = new();

    public Preferences Clone() => new()
    {
        DownloadDir = DownloadDir,
        AskDestination = AskDestination,
        Concurrency = Concurrency,
        DefaultTracks = [.. DefaultTracks],
        Proxy = new ProxySettings { Mode = Proxy.Mode, Url = Proxy.Url },
    };
}

public sealed class SettingsInfo
{
    public Preferences Preferences { get; set; } = new();
    public string DefaultDownloadDir { get; set; } = "";
    public int ConcurrencyMax { get; set; } = 8;
    public string DataDir { get; set; } = "";
    public string EngineVersion { get; set; } = "";
    public bool TestMode { get; set; }
    public bool FakeSchool { get; set; }
}

public sealed class InitializeResult
{
    public int Protocol { get; set; }
    public string Version { get; set; } = "";
    public string DataDir { get; set; } = "";
    public SettingsInfo Settings { get; set; } = new();
    public AccountInfo Account { get; set; } = new();
}
